File: jput-core/src/main/java/cz/encircled/jput/trend/PerformanceTrend.java
package cz.encircled.jput.trend;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * @author Vlad on 27-May-17.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target({ ElementType.METHOD })
public @interface PerformanceTrend {

    /**
     * Static average time threshold.
     * <p>
     *  Performance trend test will fail if average execution time is greater than average time of standard sample plus given threshold
     * </p>
     */
    double averageTimeThreshold() default -1;

    /**
     * true - use statistic variance of base sample as an average time threshold
     * <p>
     *  Performance trend test will fail if average execution time is greater than average time of standard sample plus its variance
     * </p>
     */
    boolean averageTimeVarianceThreshold() default false;

    PercentileThreshold[] percentiles() default {};

}

File: jput-core/src/main/java/cz/encircled/jput/trend/PerformanceTrend.kt
package cz.encircled.jput.trend

/**
 * @author Vlad on 27-May-17.
 */
@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class PerformanceTrend(

        /**
         * Sample size which is used for trend analysis,
         * i.e. only specified count of previous test executions will be used for analysis
         */
        val sampleSize: Int = 30,

        /**
         * Defines the way, how the sample (a subset of results) should be chosen from the all available previous results
         */
        val sampleSelectionStrategy: SelectionStrategy,

        /**
         * Static average time threshold.
         *
         * Performance trend test will fail if average execution time is greater than sample average time plus given threshold
         */
        val averageTimeThreshold: Double = 0.0,

        /**
         * if true - use the sample standard deviation as an average time threshold
         *
         * Performance trend test will fail if average execution time is greater than sample average time + threshold
         */
        val useStandardDeviationAsThreshold: Boolean = false

)